Index: py-toshi-client/tests/test_integration.py
===================================================================
diff --git a/py-toshi-client/tests/test_integration.py b/tests/test_integration.py
rename from py-toshi-client/tests/test_integration.py
rename to tests/test_integration.py
--- a/py-toshi-client/tests/test_integration.py	
+++ b/tests/test_integration.py	
@@ -4,20 +4,21 @@
 import pytest
 import requests
 
-from client import ToshiClient
-from errors import ToshiIndexError
-from index import IndexSummary, IndexSettings
-from index.field_options import TextOptionIndexing
-from index.index_builder import IndexBuilder
-from models.document import Document
-from query import TermQuery
-from query.bool_query import BoolQuery
-from query.facet_query import FacetQuery
-from query.fuzzy_query import FuzzyQuery
-from query.phrase_query import PhraseQuery
-from query.range_query import RangeQuery
-from query.regex_query import RegexQuery
 from tests.conftest import CI
+from toshi_client.client import ToshiClient
+from toshi_client.errors import ToshiIndexError
+from toshi_client.index import IndexBuilder, IndexSummary, IndexSettings
+from toshi_client.index.field_options import TextOptionIndexing
+from toshi_client.models.document import Document
+from toshi_client.query import (
+    TermQuery,
+    FuzzyQuery,
+    RangeQuery,
+    BoolQuery,
+    RegexQuery,
+    PhraseQuery,
+    FacetQuery,
+)
 
 
 class Lyrics(Document):
